### ISSUE TITLE
Kadabra 0.3.3: Pigeon FCM stops working

### ISSUE THREAD

**Issue by talklittle** [codedge-llc/kadabra#22]
I think Kadabra 0.3.3 introduced a regression where Pigeon FCM stops working after the socket times out. The push notifications do work initially after cold launching the application, but after several hours the push notifications are no longer going out.

Strangely I'm not seeing any errors from Pigeon even though I have `config :pigeon, debug_log: true`.

**Comment by hpopp** [comment]
I wonder if FCM servers changed their behavior. I'm not getting the same GOAWAYs that normally result after a couple minutes.

Edit: Nevermind, it eventually closed. I'll keep a watch throughout the day and check behavior.

**Comment by hpopp** [comment]
Haven't seen anything yet with connections stalling out. What kind of volume of pushes were you trying to send when it stopped working?

**Comment by talklittle** [comment]
It was around 5 pushes per second. The stalling happened 3 different times, after initially working for a few hours each time I restarted the application, so that led me to believe it's related to the timeouts/reconnections.

Kadabra 0.3.2 is still working normally for me.

**Comment by hpopp** [comment]
Discovered a (probably related) regression. Connections weren't properly respecting settings updates. If trying to send more than 100 FCM pushes at any given moment, anything beyond that limit will fail, and the whole batch probably fails as a result.

I'll have an 0.3.4 release out shortly, plus I'll need to update pigeon to require the new patch.

**Comment by talklittle** [comment]
Thanks. Unfortunately Pigeon 1.1.1 / Kadabra 0.3.4 still disconnected from FCM after 1 hour from cold start.

**Comment by hpopp** [comment]
The changes in [v0.3.3](https://github.com/codedge-llc/kadabra/commit/26a1e73a4b1174db1552f49890866d8ffc436719) were relatively small:
- Hpack gets a `close/1` for stopping encoders/decoders
- Streams get a `close/2` for stopping
- Connection makes theses calls on send GOAWAY (user close), recv GOAWAY (server close), and regular SSL socket close

Is there something in your codebase that might be interfering with this behavior?

Apart from that, the only other two things I can think of:
1. FCM may need a ping period, if it isn't set up already. APNS uses it because connections will silently drop otherwise.
2. Something related to this:
```elixir
 def maybe_reconnect(%{reconnect: false, client: pid} = state) do
    Logger.debug "Socket closed, not reopening, informing client"
    send(pid, {:closed, self()})
    close(state)
    {:noreply, reset_state(state, nil)}
  end
```

That block *should* be what gets invoked when the socket closes, informing the client pid that it has closed. Regardless of whether or not its the issue, `{:noreply, reset_state(state, nil)}` might be a memory leak in itself. I'll need to take a look.

**Comment by talklittle** [comment]
Here is a crash log. Sorry for not posting one much sooner:

```
18:43:04.775 [error] Got GOAWAY, NO_ERROR, Last Stream: 2147483647, Rest: max_age
18:43:04.785 [error] GenServer #PID<0.1686.0> terminating
** (stop) no process: the process is not alive or there's no process currently associated with the given name, possibly because its application isn't started
    (stdlib) gen.erl:279: :gen.do_for_proc/2
    (kadabra) lib/connection.ex:149: anonymous fn/3 in Kadabra.Connection.close/1
    (elixir) lib/enum.ex:1839: anonymous fn/3 in Enum.reduce/3
    (elixir) lib/enum.ex:3176: Enumerable.List.reduce/3
    (elixir) lib/enum.ex:1838: Enum.reduce/3
    (kadabra) lib/connection.ex:148: Kadabra.Connection.close/1
    (kadabra) lib/connection.ex:195: Kadabra.Connection.recv/2
    (stdlib) gen_server.erl:616: :gen_server.try_dispatch/4
Last message: {:"$gen_cast", {:recv, %Kadabra.Frame.Goaway{debug_data: "max_age", error_code: 0, last_stream_id: 2147483647}}}

=CRASH REPORT==== 1-Nov-2017::18:43:04 ===
  crasher:
    initial call: Elixir.Kadabra.Connection:init/1
    pid: <0.1686.0>
    registered_name: []
    exception exit: noproc
      in function  gen:do_for_proc/2 (gen.erl, line 279)
      in call from 'Elixir.Kadabra.Connection':'-close/1-fun-0-'/3 (lib/connection.ex, line 149)
      in call from 'Elixir.Enum':'-reduce/3-fun-0-'/3 (lib/enum.ex, line 1839)
      in call from 'Elixir.Enumerable.List':reduce/3 (lib/enum.ex, line 3176)
      in call from 'Elixir.Enum':reduce/3 (lib/enum.ex, line 1838)
      in call from 'Elixir.Kadabra.Connection':close/1 (lib/connection.ex, line 148)
      in call from 'Elixir.Kadabra.Connection':recv/2 (lib/connection.ex, line 195)
      in call from gen_server:try_dispatch/4 (gen_server.erl, line 616)
    ancestors: [fcm_default,pigeon,<0.1682.0>]
    message_queue_len: 1
    messages: [{finished,#{'__struct__' => 'Elixir.Kadabra.Stream.Response',
                              body =>
                                  <<"{\"multicast_id\":snip,\"success\":2,\"failure\":0,\"canonical_ids\":0,\"results\":[{\"message_id\":\"0:snip%snip\"},{\"message_id\":\"0
:snip%snip\"}]}">>,
                              headers =>
                                  [{<<":status">>,<<"200">>},
                                   {<<"content-type">>,
                                    <<"application/json; charset=UTF-8">>},
                                   {<<"date">>,
                                    <<"Wed, 01 Nov 2017 18:43:04 GMT">>},
                                   {<<"expires">>,
                                    <<"Wed, 01 Nov 2017 18:43:04 GMT">>},
                                   {<<"cache-control">>,
                                    <<"private, max-age=0">>},
                                   {<<"x-content-type-options">>,
                                    <<"nosniff">>},
                                   {<<"x-frame-options">>,<<"SAMEORIGIN">>},
                                   {<<"x-xss-protection">>,
                                    <<"1; mode=block">>},
                                   {<<"server">>,<<"GSE">>},
                                   {<<"alt-svc">>,
                                    <<"quic=\":443\"; ma=2592000; v=\"41,39,38,37,35\"">>},
                                   {<<"accept-ranges">>,<<"none">>},
                                   {<<"vary">>,<<"Accept-Encoding">>}],
                              id => 7609,status => 200}}]
    links: [<0.1685.0>]
    dictionary: []
    trap_exit: false
    status: running
    heap_size: 1598
    stack_size: 27
    reductions: 1825222
  neighbours:
    neighbour:
      pid: <0.1685.0>
      registered_name: fcm_default
      initial call: Elixir.Pigeon.Worker:init/1
      current_function: {gen_server,loop,7}
      ancestors: [pigeon,<0.1682.0>]
      message_queue_len: 0
      links: [<0.1683.0>,<0.1686.0>]
      trap_exit: false
      status: waiting
      heap_size: 4185
      stack_size: 10
      reductions: 8968235
      current_stacktrace: [{gen_server,loop,7,[{file,"gen_server.erl"},{line,381}]},
                  {proc_lib,init_p_do_apply,3,
                            [{file,"proc_lib.erl"},{line,247}]}]

=SUPERVISOR REPORT==== 1-Nov-2017::18:43:04 ===
     Supervisor: {local,pigeon}
     Context:    child_terminated
     Reason:     noproc
     Offender:   [{pid,<0.1685.0>},
                  {id,fcm_default},
                  {mfargs,
                      {'Elixir.Pigeon.Worker',start_link,
                          [#{'__struct__' => 'Elixir.Pigeon.FCM.Config',
                             key =>
                                 <<"snip">>,
                             name => fcm_default,port => 443,
                             uri => "fcm.googleapis.com"}]}},
                  {restart_type,temporary},
                  {shutdown,5000},
                  {child_type,worker}]

```

**Comment by hpopp** [comment]
Heh wow, the connection exhausted the available number of stream identifiers. The crash is happening because it appears to be trying to close stream IDs that don't exist. This is actually a fix for both Kadabra and Pigeon, as I'll need to rework the internal API for stream ID handling.

If you need a fix quickly, I'd recommend round-robining several FCM workers and periodically killing them.

**Comment by talklittle** [comment]
This appears fixed in Kadabra 0.3.5 and Pigeon 1.1.2. Thank you @hpopp
